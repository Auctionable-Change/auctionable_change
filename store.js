import React, { createContext, useContext, useReducer } from "react";

const StoreContext = createContext();

const initialState = {
  currentListing: {
    id: null,
    title: "",
    description: "",
    price: null,
    donor: "",
    donor_email: "",
    status: "",
    category: "",
    charity: "",
    charity_url: "",
    charity_score: "",
    charity_score_image: "",
    image: "",
    bids: [],
  },
  buyerDetails: {
    item_id: null,
    bidder_name: "",
    bidder_email: "",
    amount: null,
    street_address: "",
    city: "",
    state: "",
    zip_code: "",
    receipt: "",
  },
  listingToPost: {},
};

const reducer = (state, action) => {
  switch (action.type) {
    case "ADD_CURRENT_LISTING":
      return {
        ...state,
        currentListing: action.currentListing,
      };
    case "ADD_SELECTED_CHARITY":
      return {
        ...state,
        listingToPost: {
          ...state.listingToPost,
          charity: action.selectedCharity.name,
          charity_url: action.selectedCharity.url,
          charity_score: action.selectedCharity.rating,
          charity_score_image: action.selectedCharity.rating_image,
        },
      };
    case "ADD_BUYER_DETAILS":
      return {
        ...state,
        buyerDetails: { ...state.buyerDetails, ...action.buyerDetails },
      };
    case "ADD_TO_LISTING":
      return {
        ...state,
        listingToPost: { ...state.listingToPost, ...action.listingToPost },
      };
    case "POST_BID":
      return {
        ...state,
        currentBid: action.bid
      };
      case "SET_USER":
      return {
        ...state,
        loginDetails: action.loginDetails,
        userInfo: action.userInfo,
        registrationDetails: action.registration
      }

    default:
      throw new Error(`Unhandled action type: ${action.type}`);
  }
};

export const StoreProvider = ({ children }) => {
  const [state, dispatch] = useReducer(reducer, initialState);

  return (
    <StoreContext.Provider value={{ state, dispatch }}>
      {children}
    </StoreContext.Provider>
  );
};

export const useStore = () => useContext(StoreContext);
