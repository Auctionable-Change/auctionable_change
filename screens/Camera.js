import React, { useState, useEffect } from "react";
import * as ImagePicker from "expo-image-picker";
import Constants from "expo-constants";
import * as Permissions from "expo-permissions";
import { cloudinaryPost } from "./apiCalls";
import { useStore } from "../store";
import { TouchableWithoutFeedback } from "react-native-gesture-handler";
import { Image, View } from "react-native";
import { Button, Text } from "native-base";
import { StyleSheet } from "react-native";
import { useNavigation } from "@react-navigation/native";
import { SafeAreaView } from "react-native-safe-area-context";

const Camera = ({ cameraType, user, prompt, title }) => {
  const [image, uploadImage] = useState(null);
  const [imageObj, setImageObj] = useState(null);
  const { dispatch } = useStore();
  const navigation = useNavigation();

  useEffect(() => {
    getPermissionAsync();
  });

  const submitHandler = async () => {
    if (imageObj) {
      await cloudinaryPost(imageObj, updatePhotoInStore);
    }
  };

  const updatePhotoInStore = (photoData) => {
    if (user === "seller") {
      dispatch({
        type: "ADD_TO_LISTING",
        listingToPost: { image: photoData },
      });
      navigation.navigate("Choose Charity");
    }
    if (user === "buyer") {
      dispatch({
        type: "ADD_BUYER_DETAILS",
        buyerDetails: { receipt: photoData },
      });
      navigation.navigate("Email");
    }
  };

  const getPermissionAsync = async () => {
    if (Constants.platform.ios) {
      const { status } = await Permissions.askAsync(Permissions.CAMERA_ROLL);
      if (status !== "granted") {
        alert("Sorry, we need camera roll permissions to make this work!");
      }
    }
    if (Constants.platform.ios) {
      const { status } = await Permissions.askAsync(Permissions.CAMERA);
      if (status !== "granted") {
        alert("Sorry, we need camera roll permissions to make this work!");
      }
    }
  };

  const _pickImage = async () => {
    try {
      let result = await ImagePicker[cameraType]({
        mediaTypes: ImagePicker.MediaTypeOptions.Image,
        allowsEditing: true,
        aspect: [4, 3],
        quality: 1,
        base64: true,
      });
      if (!result.cancelled) {
        uploadImage(result.uri);
        setImageObj(result);
      }
    } catch (E) {
      console.log(E);
    }
  };
  return (
    <SafeAreaView style={styles.mainContainer}>
      <Text style={styles.title}>{title}</Text>
      <View style={styles.container}>
        <Text accessibilityLabel={prompt} color="#2cb833" onPress={_pickImage}>
          {prompt}
        </Text>
        {image ? (
          <Image source={{ uri: image }} style={{ width: 200, height: 200 }} />
        ) : (
          <TouchableWithoutFeedback onPress={_pickImage}>
            <Image
              source={require("../assets/icons/camera.png")}
              style={{ width: 150, height: 150 }}
            />
          </TouchableWithoutFeedback>
        )}
        <Button block success onPress={() => submitHandler()}>
          <Text>Continue</Text>
        </Button>
      </View>
    </SafeAreaView>
  );
};

const styles = StyleSheet.create({
  mainContainer: {
    flex: 1,
    alignItems: "center",
    padding: 20,
  },
  container: {
    alignItems: "center",
    backgroundColor: "white",
    padding: 20,
    borderRadius: 10,
    shadowOffset: {
      width: 0,
      height: 1,
    },
    shadowOpacity: 0.18,
    shadowRadius: 1.0,
  },
  title: {
    fontSize: 25,
    textAlign: "center",
    marginBottom: 20,
  },
});

export default Camera;
