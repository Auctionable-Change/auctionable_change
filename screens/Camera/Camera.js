import React, { useState, useEffect } from "react";
import * as ImagePicker from "expo-image-picker";
import Constants from "expo-constants";
import * as Permissions from "expo-permissions";
import { cloudinaryPost } from "../apiCalls";
import { useStore } from "../../store";
import { TouchableWithoutFeedback } from "react-native-gesture-handler";
import { Alert, Image } from "react-native";
import { Button, Text } from "native-base";
import { StyleSheet } from "react-native";
import { useNavigation } from "@react-navigation/native";
import { SafeAreaView } from "react-native-safe-area-context";

const Camera = ({ cameraType, user, prompt, title }) => {
  const [image, uploadImage] = useState(null);
  const [imageObj, setImageObj] = useState(null);
  const { dispatch } = useStore();
  const navigation = useNavigation();

  useEffect(() => {
    getPermissionAsync();
  });

  const submitHandler = async () => {
    if (imageObj) {
      await cloudinaryPost(imageObj, updatePhotoInStore);
    } else {
      Alert.alert("Photo Required", "You must upload photo to continue");
    }
  };

  const updatePhotoInStore = (photoData) => {
    if (user === "seller") {
      dispatch({
        type: "ADD_TO_LISTING",
        listingToPost: { image: photoData },
      });
      navigation.navigate("Choose Charity");
    }
    if (user === "buyer") {
      dispatch({
        type: "ADD_BUYER_DETAILS",
        buyerDetails: { receipt: photoData },
      });
      navigation.navigate("Email");
    }
  };

  const getPermissionAsync = async () => {
    if (Constants.platform.ios) {
      const { status } = await Permissions.askAsync(Permissions.CAMERA_ROLL);
      if (status !== "granted") {
        alert("Sorry, we need camera roll permissions to make this work!");
      }
    }
    if (Constants.platform.ios) {
      const { status } = await Permissions.askAsync(Permissions.CAMERA);
      if (status !== "granted") {
        alert("Sorry, we need camera roll permissions to make this work!");
      }
    }
  };

  const _pickImage = async () => {
    try {
      let result = await ImagePicker[cameraType]({
        mediaTypes: ImagePicker.MediaTypeOptions.Image,
        allowsEditing: true,
        aspect: [4, 3],
        quality: 1,
        base64: true,
      });
      if (!result.cancelled) {
        uploadImage(result.uri);
        setImageObj(result);
      }
    } catch (E) {
      console.log(E);
    }
  };
  return (
    <SafeAreaView style={styles.container}>
      <Text style={styles.title}>{title}</Text>
      <Text
        accessibilityLabel={prompt}
        color="#2cb833"
        onPress={_pickImage}
        style={{
          fontFamily: "quicksand",
          fontSize: 15,
          margin: 10,
          textAlign: "center",
        }}
      >
        {prompt}
      </Text>
      {image ? (
        <Image source={{ uri: image }} style={styles.image} />
      ) : (
        <TouchableWithoutFeedback onPress={_pickImage}>
          <Image
            source={require("../../assets/icons/ar-camera.png")}
            style={styles.camera}
          />
        </TouchableWithoutFeedback>
      )}
      <Button
        rounded
        success
        onPress={() => submitHandler()}
        style={styles.button}
      >
        <Text style={{ fontFamily: "quicksand-bold" }}>Continue</Text>
      </Button>
    </SafeAreaView>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: "center",
    backgroundColor: "#FFFFFF",
  },
  title: {
    fontSize: 20,
    textAlign: "center",
    marginBottom: 10,
    fontFamily: "quicksand-bold",
  },
  image: { margin: 5, width: 300, height: 300 },
  camera: { width: 150, height: 150 },
  button: { alignSelf: "center", backgroundColor: "#2cb833", margin: 5 },
});

export default Camera;
