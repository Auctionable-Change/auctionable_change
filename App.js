import React, { useEffect } from "react";
import { AppLoading } from "expo";
import * as Font from "expo-font";
import { Ionicons } from "@expo/vector-icons";
import { createStackNavigator } from "@react-navigation/stack";
import { NavigationContainer } from "@react-navigation/native";
import { StoreProvider } from "./store";
import { StyleSheet } from "react-native";
import LogoTitle from "./components/LogoTitle";
import Welcome from "./screens/Welcome/Welcome";
import CurrentListings from "./screens/CurrentListings/CurrentListings";
import ListingDetails from "./screens/ListingDetails/ListingDetails";
import PostItem from "./screens/PostItem/PostItem";
import PaymentInstructions from "./screens/PaymentInstructions/PaymentInstructions";
import ChooseCharity from "./screens/ChooseCharity/ChooseCharity";
import PostConfirmation from "./screens/PostConfirmation/PostConfirmation";
import EmailForm from "./screens/EmailForm/EmailForm";
import Camera from "./screens/Camera/Camera";
import { Image } from "react-native";

const Stack = createStackNavigator();

export default function App() {
  useEffect(() => {
    const loadFonts = async () => {
      await Font.loadAsync({
        Roboto: require("native-base/Fonts/Roboto.ttf"),
        Roboto_medium: require("native-base/Fonts/Roboto_medium.ttf"),
        ...Ionicons.font,
      });
    };
    loadFonts();
  });

  return (
    <StoreProvider>
      <NavigationContainer>
        <Stack.Navigator>
          <Stack.Screen name="Home" component={Welcome} />
          <Stack.Screen
            name="Listings"
            component={CurrentListings}
            options={{
              headerTitle: () => <LogoTitle />,
              headerBackTitleVisible: false,
              ...backButtonStyle,
            }}
          />
          <Stack.Screen
            name="Details"
            component={ListingDetails}
            options={{
              headerTitle: () => <LogoTitle />,
              headerBackTitleVisible: false,
              ...backButtonStyle,
            }}
          />
          <Stack.Screen
            name="Purchase"
            component={PaymentInstructions}
            options={{
              headerTitle: () => <LogoTitle />,
              headerBackTitleVisible: false,
              ...backButtonStyle,
            }}
          />
          <Stack.Screen
            name="Email"
            component={EmailForm}
            options={{
              headerTitle: () => <LogoTitle />,
              headerBackTitleVisible: false,
              ...backButtonStyle,
            }}
          />
          <Stack.Screen
            name="Post Item"
            component={PostItem}
            options={{
              headerTitle: () => <LogoTitle />,
              headerBackTitleVisible: false,
              ...backButtonStyle,
            }}
          />
          <Stack.Screen
            name="Choose Charity"
            component={ChooseCharity}
            options={{
              headerTitle: () => <LogoTitle />,
              headerBackTitleVisible: false,
              ...backButtonStyle,
            }}
          />
          <Stack.Screen
            name="Confirmation"
            component={PostConfirmation}
            options={{
              headerTitle: () => <LogoTitle />,
              headerBackTitleVisible: false,
              ...backButtonStyle,
            }}
          />
          <Stack.Screen
            name="Camera"
            options={{
              headerTitle: () => <LogoTitle />,
              headerBackTitleVisible: false,
              ...backButtonStyle,
            }}
          >
            {() => <Camera cameraType="launchCameraAsync"
              user="seller"
              prompt="Press camera to take a photo of your item to donate!"
              title="Upload a Photo"/>}
          </Stack.Screen>
        </Stack.Navigator>
      </NavigationContainer>
    </StoreProvider>
  );
}

const backButtonStyle = {
  headerBackImage: () => (
    <Image
      accessibilityRole="button"
      id="add"
      source={require("./assets/icons/back.png")}
      style={{ width: 30, height: 30, margin: 8 }}
    />
  )         
}