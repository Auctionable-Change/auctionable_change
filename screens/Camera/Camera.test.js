import React from 'react';
import { render, fireEvent, waitFor, NativeTestEvent } from 'react-native-testing-library';
import Camera from './Camera';
import App from '../../App'
import { StoreProvider } from '../../store'
import {expect, it} from '@jest/globals';
import renderer from "react-test-renderer";
import { NavigationContainer } from '@react-navigation/native'
import { debug } from 'react-native-reanimated';


describe('Camera', () => {
 it('should render', () => {
    const tree = renderer.create(
          <App component={Camera}/>
      ).toJSON();
    expect(tree).toMatchSnapshot();
  })

  // it('should display display a Continue Button on load', () => {
  //   const { getByText, getByA11yLabel, debug } = render(
  //     <StoreProvider>
  //        <App component={Camera}
  //          params={{title: 'test'}}
  //        />
  //     </StoreProvider>
  //   )
  //   debug()
  //   expect(getByText('test')).toBeTruthy()
  // })
})