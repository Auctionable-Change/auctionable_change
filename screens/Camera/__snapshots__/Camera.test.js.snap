// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Camera should render 1`] = `null`;
